Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp

```diff
@@ -892,29 +892,39 @@ void GraphicsLayerCoordinated::updateVisibleRect(const FloatRect& rect)
 }
 
 #if ENABLE(DAMAGE_TRACKING)
-void GraphicsLayerCoordinated::updateDamage()
+void GraphicsLayerCoordinated::updateDamage(const FloatSize& visibleSize)
 {
     if (!m_platformLayer->damagePropagation())
         return;
 
-    Damage damage;
-    if (m_dirtyRegion.fullRepaint) {
-        // FIXME: this will be simply changing the mode when we create damage with the layer size.
-        damage.resize(m_size);
-        damage.setMode(Damage::Mode::Full);
-        damage.add(FloatRect({ }, m_size));
-    } else {
+    auto damageSize = [&]() -> FloatSize {
+        if (!m_dirtyRegion.fullRepaint) {
+            static constexpr float minDamageSize = 512;
+            if (m_size.width() < minDamageSize || m_size.height() < minDamageSize)
+                return { minDamageSize, minDamageSize };
+
+            if (m_size.width() > visibleSize.width() || m_size.height() > visibleSize.height())
+                return visibleSize;
```

Comment:
IMO in case of e.g. 3000x512 layer and 2000x1000 viewport, the above should return 2000x512
similarly, IMO in case of e.g. 1000x256 layer and 2000x1000 viewport, the above should return 1000x512

so in other words, we should:
```
return {
min(max(m_size.width(), 512), visibleSize.width()),
min(max(m_size.height(), 512), visibleSize.height())
}
```